src/D_functions_and_procedures.sql

```diff
@@ -21,4 +21,89 @@ $$
 LANGUAGE 'plpgsql';
 
 --D2 TEST
-SELECT * FROM simple_search('apple',1);
\ No newline at end of file
+SELECT * FROM simple_search('apple',1);
+
+
+--D3 Rating function
+CREATE OR REPLACE FUNCTION rate(p_userid INT, p_imdb_id VARCHAR, p_score NUMERIC)
+RETURNS VOID AS $$
+DECLARE
+    v_media_id INTEGER;
+    v_existing_user_score RECORD;
+BEGIN
+    -- Get media ID from IMDb ID
+    SELECT media_id INTO v_media_id FROM media WHERE imdb_id = p_imdb_id;
+
+    -- Check if the user has already rated this media in user_score
+    SELECT * INTO v_existing_user_score 
+    FROM user_score 
+    WHERE user_id = p_userid AND media_id = v_media_id;
+
+    IF FOUND THEN
```

Comment:
what is this magic 👀

Reply:
Magic 🪄 : 
IF FOUND THEN is shorthand that checks whether the previous SELECT statement found any data.

If a row was found (i.e., the user has already rated the movie), the block inside IF FOUND THEN will execute.
If no row was found, it means the user has never rated the movie, and the ELSE block will run.

Reply:
That is pretty cool